ui/desktop/src/components/bottom_menu/DirSwitcher.tsx

```diff
@@ -1,23 +1,84 @@
-import React, { useState } from 'react';
+import React, { useState, useEffect } from 'react';
 import { FolderDot } from 'lucide-react';
 import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from '../ui/Tooltip';
+import { updateSessionWorkingDir, restartAgent, getSession } from '../../api';
+import { toast } from 'react-toastify';
+import { setWorkingDir, getWorkingDir } from '../../store/newChatState';
 
 interface DirSwitcherProps {
-  className?: string;
+  className: string;
+  sessionId: string | undefined;
 }
 
-export const DirSwitcher: React.FC<DirSwitcherProps> = ({ className = '' }) => {
+export const DirSwitcher: React.FC<DirSwitcherProps> = ({ className, sessionId }) => {
   const [isTooltipOpen, setIsTooltipOpen] = useState(false);
   const [isDirectoryChooserOpen, setIsDirectoryChooserOpen] = useState(false);
 
+  const [sessionWorkingDir, setSessionWorkingDir] = useState<string | null>(null);
+
+  // Fetch the working directory from the session when sessionId changes
+  useEffect(() => {
+    if (!sessionId) {
+      setSessionWorkingDir(null);
+      return;
+    }
+
+    const fetchSessionWorkingDir = async () => {
+      try {
+        const response = await getSession({ path: { session_id: sessionId } });
+        if (response.data?.working_dir) {
+          setSessionWorkingDir(response.data.working_dir);
+        }
+      } catch (error) {
+        console.error('[DirSwitcher] Failed to fetch session working dir:', error);
+      }
+    };
+
+    fetchSessionWorkingDir();
+  }, [sessionId]);
+
+  const currentDir = sessionWorkingDir ?? getWorkingDir();
+
   const handleDirectoryChange = async () => {
     if (isDirectoryChooserOpen) return;
     setIsDirectoryChooserOpen(true);
+
+    let result;
     try {
-      await window.electron.directoryChooser(true);
+      result = await window.electron.directoryChooser();
     } finally {
       setIsDirectoryChooserOpen(false);
     }
+
+    if (result.canceled || result.filePaths.length === 0) {
+      return;
+    }
+
+    const newDir = result.filePaths[0];
+    setWorkingDir(newDir);
+
+    await window.electron.addRecentDir(newDir);
+
+    if (sessionId) {
+      try {
+        await updateSessionWorkingDir({
+          path: { session_id: sessionId },
+          body: { workingDir: newDir },
+        });
+
+        try {
+          await restartAgent({ body: { session_id: sessionId } });
```

Comment:
this shouldn't be a separate call. we should have an agentUpdateWorkingDir that might also restart the agent - the client shouldn't know about this implementation detail